ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -44,15 +44,40 @@ Alignment legacyShapeToAlignment(const ttnn::Shape& shape) {
 }
 
 TensorLayout::TensorLayout(DataType dtype, const PageConfig& page_config, const MemoryConfig& memory_config)
-    : TensorLayout(dtype, page_config, memory_config, {}) {
+    : TensorLayout(dtype, page_config, memory_config, std::nullopt, {}) {
+}
+
+TensorLayout::TensorLayout(DataType dtype, const PageConfig& page_config, const MemoryConfig& memory_config, const std::optional<Size>& shard_shape, const Alignment& shard_alignment)
+    : dtype_(dtype),
+      page_config_(page_config),
+      memory_config_(memory_config),
+      alignment_(shard_alignment) {
+
+    // TODO: If we remove alignment, we should lower this into validate_alignment
+    TT_FATAL(shard_alignment.size() <= 2, "Shard alignment {} exceeds maximum rank of 2", shard_alignment);
+
+    if (shard_shape.has_value()) {
+        TT_FATAL(memory_config.is_sharded(), "Shard shape should only be used if memory config is sharded");
+        this->shard_shape_ = shard_shape.value();
+    } else if (memory_config.shard_spec.has_value()) {
+        // Treat shard shape from shard spec as physical shard shape and use here as logical shard shape for this->shard_shape_
+        // When computing ShardSpecBuffer, double check that the computed physical shard shape is the same as shard shape from shard spec
+        this->shard_shape_ = Size(memory_config.shard_spec.value().shape);
```

Comment:
I'm a bit confused by this code. Would it make sense to rename `this->shard_shape_` to `this->logical_shard_shape_`? I'm unsure if the comment clarifies why we set `shard_shape_` either to `memory_config.shard_spec.value().shape`, it seems like user is providing both...